server/environment/manager.go

```diff
@@ -0,0 +1,331 @@
+package environment
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/kubeshop/tracetest/server/pkg/id"
+	"github.com/kubeshop/tracetest/server/resourcemanager"
+)
+
+type Repository struct {
+	db *sql.DB
+}
+
+func NewRepository(db *sql.DB) Repository {
+	return Repository{db}
+}
+
+type scanner interface {
+	Scan(dest ...interface{}) error
+}
+
+const (
+	insertQuery = `
+		INSERT INTO environments (
+			"id",
+			"name",
+			"description",
+			"created_at",
+			"values"
+		) VALUES ($1, $2, $3, $4, $5)`
+
+	updateQuery = `
+		UPDATE environments SET
+			"id" = $2,
+			"name" = $3,
+			"description" = $4,
+			"created_at" = $5,
+			"values" = $6
+		WHERE id = $1
+	`
+
+	getQuery = `
+	SELECT
+		e.id,
+		e.name,
+		e.description,
+		e.created_at,
+		e.values
+	FROM environments e
+`
+
+	deleteQuery = "DELETE FROM environments WHERE id = $1"
+
+	idExistsQuery = `
+		SELECT COUNT(*) > 0 as exists FROM environments WHERE id = $1
+	`
+
+	countQuery = `
+		SELECT COUNT(*) FROM environments e
+	`
+)
+
+var _ resourcemanager.Create[Environment] = &Repository{}
+var _ resourcemanager.Delete[Environment] = &Repository{}
+var _ resourcemanager.Get[Environment] = &Repository{}
+var _ resourcemanager.List[Environment] = &Repository{}
+var _ resourcemanager.Update[Environment] = &Repository{}
+var _ resourcemanager.IDSetter[Environment] = &Repository{}
+var _ resourcemanager.Provision[Environment] = &Repository{}
+
+func (*Repository) SortingFields() []string {
+	return []string{"name", "createdAt"}
+}
+
+func (r *Repository) SetID(environment Environment, id id.ID) Environment {
+	environment.ID = id
+	return environment
+}
+
+func (r *Repository) Create(ctx context.Context, environment Environment) (Environment, error) {
+	environment.ID = environment.Slug()
+	environment.CreatedAt = time.Now().UTC().Format(time.RFC3339Nano)
+
+	return r.insertIntoEnvironments(ctx, environment)
+}
+
+func (r *Repository) Update(ctx context.Context, environment Environment) (Environment, error) {
+	oldEnvironment, err := r.Get(ctx, environment.ID)
+	if err != nil {
+		return Environment{}, err
+	}
+
+	// keep the same creation date to keep sort order
+	environment.CreatedAt = oldEnvironment.CreatedAt
+	environment.ID = oldEnvironment.ID
+
+	return r.updateIntoEnvironments(ctx, environment, oldEnvironment.ID)
+}
+
+func (r *Repository) Delete(ctx context.Context, id id.ID) error {
+	_, err := r.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	tx, err := r.db.BeginTx(ctx, nil)
+	if err != nil {
+		return fmt.Errorf("sql BeginTx: %w", err)
+	}
+
+	_, err = tx.ExecContext(ctx, deleteQuery, id)
+
+	if err != nil {
+		tx.Rollback()
+		return fmt.Errorf("sql error: %w", err)
+	}
```

Comment:
We can defer a rollback function here:

```suggestion
	tx, err := r.db.BeginTx(ctx, nil)
	if err != nil {
		return fmt.Errorf("sql BeginTx: %w", err)
	}
	defer tx.Rollback()

	_, err = tx.ExecContext(ctx, deleteQuery, id)
	if err != nil {
		return fmt.Errorf("sql error: %w", err)
	}
```